### ISSUE TITLE
[FEATURE REQUEST] Ability to make tags

### ISSUE THREAD

**Issue by hhsecond** [tensorwerk/hangar-py#157]
**Is your feature request related to a problem? Please describe.**
Ability to create tags and checkout to tags is important especially when we try to make a wrapper like the stockroom

**Describe the solution you'd like**
After commit, like GitHub, the user should be able to tag. User should also be able to checkout to tags

**Comment by elistevens** [comment]
FYI, I think that in the short term, branches can accomplish most of the needed functionality.

**Comment by hhsecond** [comment]
Agreed, but adding tags is not a difficult task either. I see why'd you think that way but I had a need for this in the wrapper I was writing for hangar for model storage. Specifying commits (tags) over branches is bit more convenient there

**Comment by rlizzo** [comment]
So Sherin, I actually think this is one feature of git which I'd prefer not to emulate. Tags were added as kind of an afterthought, and have a couple really horrifying properties:

- Since there's nothing connecting them directly to the immutable history of the project, they don't offer any sort of guarantee there's something tagged version x.Y.Z will remain that exact same tag if the project to maintainer where to go a little crazy and remove or replace it on another commit. 

- Because there's nothing connecting a tag with the rest of the ancestor graph, the way in which they are transferred between clients and servers is not deterministic. Were a tag Incorporated directly into the merkle tree, it would be possible to automatically push and pull tags without trust (history can be independently verified). Since that's not the case, fetching and pushing tags has remained this orphaned operation in Git. 

At the end of the day what we are looking for is a away to assign a human readable name to a particular commit digest / point in time. I agree with Eli that branches solve most of this right now, and that we've got bigger things to work on. I know keeping the digest around isn't the most convenient thing in the world, but it really is the best way to refer to a commit. 

I'm sure we could figure out a solution that would actually work here for a project like this at its stage in its life cycle. We could technically add the concept of a tag field in the components that are hashed along each commit, or we might even come up with something a bit more clever. _I don't want to discourage you from working on this if it's something that you need, but getting to an acceptable solution isn't really as trivial as it appears at first._
